pkg/controller/cloudlet_api.go

```diff
@@ -2626,6 +2653,133 @@ func (s *CloudletApi) updateDefaultMultiTenantClusterWorker(ctx context.Context,
+				return nil
+			}
+			zoneKey := cloudlet.GetZone()
+			if ai.ZoneKey.Matches(zoneKey) {
+				// no change
+				return nil
+			}
+			ai.ZoneKey = *zoneKey
+			s.all.appInstApi.store.STMPut(stm, &ai)
+			return nil
+		})
+		if err != nil {
+			log.SpanLog(ctx, log.DebugLevelApi, "failed to update zone for app inst", "appinst", aikey, "err", err)
+		}
+	}
+	ciFilter := edgeproto.ClusterInst{
+		CloudletKey: *ckey,
+	}
+	ciKeys := []*edgeproto.ClusterKey{}
+	s.all.clusterInstApi.cache.Show(&ciFilter, func(obj *edgeproto.ClusterInst) error {
+		ciKeys = append(ciKeys, &obj.Key)
+		return nil
+	})
+	log.SpanLog(ctx, log.DebugLevelApi, "update zone for clusterinsts", "cloudlet", ckey)
+	for _, cikey := range ciKeys {
+		err := s.all.clusterInstApi.sync.ApplySTMWait(ctx, func(stm concurrency.STM) error {
+			ci := edgeproto.ClusterInst{}
+			if !s.all.clusterInstApi.store.STMGet(stm, cikey, &ci) {
+				return nil
+			}
+			if !ci.CloudletKey.Matches(ckey) {
+				return nil
+			}
+			cloudlet := edgeproto.Cloudlet{}
+			if !s.all.cloudletApi.store.STMGet(stm, ckey, &cloudlet) {
+				return nil
+			}
+			zoneKey := cloudlet.GetZone()
+			if ci.ZoneKey.Matches(zoneKey) {
+				// no change
+				return nil
+			}
+			ci.ZoneKey = *zoneKey
+			s.all.clusterInstApi.store.STMPut(stm, &ci)
+			return nil
+		})
+		if err != nil {
+			log.SpanLog(ctx, log.DebugLevelApi, "failed to update zone for cluster inst", "clusterinst", cikey, "err", err)
+		}
+	}
+}
+
+func (s *CloudletApi) updateZoneLocation(ctx context.Context, org string, oldZone, newZone string) {
+	filter := edgeproto.Cloudlet{}
+	filter.Key.Organization = org
+	locations := map[edgeproto.ZoneKey][]dme.Loc{}
+	// We need to potentially update two zones, the zone the cloudlet used to
+	// be assigned to, and the new zone the cloudlet is now assigned
+	err := s.cache.Show(&filter, func(cloudlet *edgeproto.Cloudlet) error {
+		if cloudlet.Zone != "" && (cloudlet.Zone == oldZone || cloudlet.Zone == newZone) {
+			zoneKey := cloudlet.GetZone()
+			locations[*zoneKey] = append(locations[*zoneKey], cloudlet.Location)
+		}
+		return nil
+	})
+	if err != nil {
+		log.SpanLog(ctx, log.DebugLevelApi, "failed to update zone location", "operator", org, "err", err)
+	}
+	for _, zoneName := range []string{oldZone, newZone} {
+		zkey := edgeproto.ZoneKey{
+			Name:         zoneName,
+			Organization: org,
+		}
+		err := s.all.zoneApi.sync.ApplySTMWait(ctx, func(stm concurrency.STM) error {
+			zone := edgeproto.Zone{}
+			if !s.all.zoneApi.store.STMGet(stm, &zkey, &zone) {
+				return nil
+			}
+			if zone.DeletePrepare {
+				return nil
+			}
+			locs, ok := locations[zkey]
+			if !ok {
+				zone.Location = dme.Loc{}
+			} else {
+				var lat, long float64
+				for _, loc := range locs {
+					lat += loc.Latitude
+					long += loc.Longitude
+				}
+				lat /= float64(len(locs))
+				long /= float64(len(locs))
+				zone.Location.Latitude = lat
+				zone.Location.Longitude = long
```

Comment:
Oh man, zone location could really be random with these calculations. Maybe specifying the location when creating a zone makes more sense? What do you think?

Reply:
That's a fair point. I think it's probably ok for now, as any point within the "zone" area is as accurate as any other for the location. But I agree we may want to let the user specify the location at some point.